kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -18,94 +18,276 @@
+        .add("protocol", Protocol.READ_SCHEMA)
+        .add("metaData", Metadata.READ_SCHEMA);
+
+    /** Read schema when searching for all the active AddFiles (need the RemoveFiles, too). */
+    public static final StructType ADD_REMOVE_READ_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA)
+        .add("remove", RemoveFile.READ_SCHEMA);
+
+    /** Data (result) schema of the remaining active AddFiles. */
+    public static final StructType ADD_ONLY_DATA_SCHEMA = new StructType()
+        .add("add", AddFile.READ_SCHEMA);
+
     private final Path dataPath;
     private final LogSegment logSegment;
-    private final CloseableIterable<Tuple2<Action, Boolean>> reverseActionsIterable;
+    private final TableClient tableClient;
+
+    /**
+     * An iterable generating iterators of ColumnarBatches of AddFiles and RemovedFiles, used to
+     * return the active AddFiles in the Snapshot.
+     */
+    private final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
+        addRemoveColumnarBatchIterable;
+
     private final Lazy<Tuple2<Protocol, Metadata>> protocolAndMetadata;
 
     public LogReplay(
-        Path logPath,
-        Path dataPath,
-        TableClient tableHelper,
-        LogSegment logSegment)
-    {
+            Path logPath,
+            Path dataPath,
+            TableClient tableClient,
+            LogSegment logSegment) {
         this.dataPath = dataPath;
         this.logSegment = logSegment;
+        this.tableClient = tableClient;
 
-        final Stream<FileStatus> allFiles = Stream.concat(
-            logSegment.checkpoints.stream(),
-            logSegment.deltas.stream());
+        final List<FileStatus> allFiles = Stream.concat(
+                logSegment.checkpoints.stream(),
+                logSegment.deltas.stream()
+            ).collect(Collectors.toList());
         assertLogFilesBelongToTable(logPath, allFiles);
 
-        this.reverseActionsIterable = new ReverseFilesToActionsIterable(
-            tableHelper,
-            allFiles);
-        this.protocolAndMetadata = new Lazy<>(this::loadTableProtocolAndMetadata);
+        this.addRemoveColumnarBatchIterable = new ReverseFilesToColumnarBatchActionsIterable(
+            tableClient,
+            allFiles,
+            ADD_REMOVE_READ_SCHEMA);
+
+        final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
+            protocolMetadataColumnarBatchIterable = new ReverseFilesToColumnarBatchActionsIterable(
+                tableClient,
+                allFiles,
+            PROTOCOL_METADATA_READ_SCHEMA);
+
+        this.protocolAndMetadata = new Lazy<>(
+            () -> loadTableProtocolAndMetadata(protocolMetadataColumnarBatchIterable)
+        );
     }
 
-    public Lazy<Tuple2<Protocol, Metadata>> lazyLoadProtocolAndMetadata()
-    {
+    /////////////////
+    // Public APIs //
+    /////////////////
+
+    public Lazy<Tuple2<Protocol, Metadata>> lazyLoadProtocolAndMetadata() {
```

Comment:
why is this lazy? whats the benefit of making this lazy? shouldnt the callers of this function decide whether to delay load it or immediately load it?

Reply:
i understand that you are defining the class field as lazy.. thats fine... its good to define these as constants right away in the constructor. but that does not need to exposed as a lazy in a public method.. does it?

Reply:
SGTM. Do you want the consumer of this (snapshot) to load this upon snapshot instantiation? or to lazily load the protocol and metadata, only upon the first access?

Reply:
load upon first access. if the consumer wants to load it async, then they can add a layer of threads and future on top of it.